src/submission/logic.py

```diff
@@ -275,15 +282,149 @@ def order_fields(request, fields):
 
 
 def save_author_order(request, article):
-    author_pks = [int(pk) for pk in request.POST.getlist('authors[]')]
-    for author in article.authors.all():
-        order = author_pks.index(author.pk)
+    author_pk = int(request.POST.get('author_pk'))
+    author = get_object_or_404(
+        core_models.Account,
+        pk=author_pk,
+    )
+    change_order = request.POST.get('change_order')
+    author_pks = [
+        order.author.pk for order in article.articleauthororder_set.all()
+    ]
+    old_order = author_pks.index(author_pk)
+    new_orders = {
+        'first': 0,
+        'up': max(old_order - 1, 0),
+        'down': min(old_order + 1, len(author_pks) - 1),
+        'last': len(author_pks) - 1,
+    }
+    new_order = new_orders[change_order]
+    if old_order == new_order:
+        return
+    author_pks.pop(old_order)
+    author_pks.insert(new_order, author_pk)
+    for i, author_pk in enumerate(author_pks):
         author_order, c = models.ArticleAuthorOrder.objects.get_or_create(
             article=article,
-            author=author,
-            defaults={'order': order}
+            author__pk=author_pk,
+            defaults={'order': i}
         )
-
         if not c:
-            author_order.order = order
+            author_order.order = i
             author_order.save()
+    sentences = {
+        'first': _('%(author_name)s moved to the start.')
+            % {'author_name': author.full_name()},
+        'up':  _('%(author_name)s moved up.')
+            % {'author_name': author.full_name()},
+        'down':  _('%(author_name)s moved down.')
+            % {'author_name': author.full_name()},
+        'last':  _('%(author_name)s moved to the end.')
+            % {'author_name': author.full_name()},
+    }
+    messages.add_message(
+        request,
+        messages.SUCCESS,
+        sentences[change_order],
+    )
+    return author
+
+
+def add_author_from_search(search_term, request, article):
+    query = Q(email=search_term)
```

Comment:
for my own sanity, can we split this function? There are just too many branches. At a first step we want to check if the search term is either an Orcid, an email or something else. Then we delegate to `add_author_via_email` and `add_author_via_orcid`

Also try to keep all updates to a record in the same area of the codebase.